File: 29-DivideTwoIntegers/29-DivideTwoIntegers.py
# Last updated: 6/22/2025, 11:05:34 AM
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        INT_MAX = 2**31 - 1
        INT_MIN = -2**31
        if dividend==INT_MIN and divisor==-1:
            return INT_MAX
        quotient=dividend/divisor
        return int(quotient)
        # if quotient<(((2**31)-1) or (-(2**31))):
        #     res=quotient-1
        #     return res

        

File: 29-DivideTwoIntegers/29-DivideTwoIntegers.java
// Last updated: 6/22/2025, 11:05:31 AM
class Solution {
    public int divide(int dividend, int divisor) {
        if (dividend == Integer.MIN_VALUE && divisor == -1) {
            return Integer.MAX_VALUE;
        }

        boolean neg = (dividend < 0) ^ (divisor < 0);
        long a = Math.abs((long) dividend);
        long b = Math.abs((long) divisor);
        int ans = 0;
        while (a >= b) {
            int count = 0;
            while (a >= (b << (count + 1))) {
                count++;
            }
            ans += 1 << count;
            a -= b << count;
        }
        
        return neg ? -ans : ans;
    }
}
